lib/irclib/client.js

```diff
@@ -13,6 +13,12 @@ var log = require("../logging").get("irc-client");
 // The length of time to wait before trying to join the channel again
 var JOIN_TIMEOUT_MS = 15 * 1000; // 15s
 
+// A global counter for generating unique numerical user IDs
+var userCount = 0;
```

Comment:
Indeed, this is in fact the trickiest part of ipv6 support. We already have to do this when generating ident usernames, so it should be possible to crib the control flow mechanics from it.

See https://github.com/matrix-org/matrix-appservice-irc/blob/master/lib/irclib/names.js for how ident username generation works - you're interested in the queueing mechanics and https://github.com/matrix-org/matrix-appservice-irc/blob/master/lib/irclib/names.js#L139 which persists the user to the DB. You will need to:
 - Add `ipv6address` to https://github.com/matrix-org/matrix-appservice-irc/blob/master/lib/models/users.js#L41
 - Add it as a persisted key in https://github.com/matrix-org/matrix-appservice-irc/blob/master/lib/store.js#L451
 - Repopulate the field on the getters - https://github.com/matrix-org/matrix-appservice-irc/blob/master/lib/store.js#L434 and https://github.com/matrix-org/matrix-appservice-irc/blob/master/lib/store.js#L488

Then you need to add the glue to make it work on each client connection:
 - The ident stuff does it here: https://github.com/matrix-org/matrix-appservice-irc/blob/4a2dc0ed0a09e746aef640fa05d30bc7534dfc8f/lib/irclib/client.js#L83
 - You would add another promise in the chain to `getIpv6Address` or something, then use it in `clientConnection.create()`

Testing wise I would expect:
 - A test which does *not* set `localAddress` if there is no ipv6 prefix supplied.
 - A test which does set `localAddress` if there is an ipv6 prefix supplied (which is correct)
 - A test which checks your queueing behaviour (sending 2 matrix messages from the same user ID to the bridge on the same tick of the event loop, should get assigned the same ipv6 address (if locks are not taken out correctly this could race))
 - A test which checks your counting (sending 2 matrix messages, one from user A, one from user B, should get sequentially higher addresses assigned.

How would the bridge respond if it consumed all the ipv6 block it had been allocated? Refuse to make new connections? Cycle them around from 0 again (how would that interact with bans on the network)?